File: android/app/src/main/kotlin/com/meshconnect/meshconnect/MeshTransportPlugin.kt
package com.meshconnect.meshconnect

import android.content.Context
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

class MeshTransportPlugin : FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler {

    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var eventSink: EventChannel.EventSink? = null
    private var context: Context? = null

    private var bleTransport: BleTransport? = null
    private var wifiTransport: WifiPeerTransport? = null
    private var permissionManager: PermissionManager? = null
    private var batteryManager: BatteryManager? = null

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext
        methodChannel = MethodChannel(flutterPluginBinding.binaryMessenger, "com.meshconnect/transport")
        methodChannel.setMethodCallHandler(this)

        eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "com.meshconnect/events")
        eventChannel.setStreamHandler(this)

        val ctx = flutterPluginBinding.applicationContext
        permissionManager = PermissionManager(ctx)
        batteryManager = BatteryManager(ctx)
        wifiTransport = WifiPeerTransport(ctx)

        bleTransport = BleTransport(
            context = ctx,
            onPeerDiscovered = { peerData ->
                eventSink?.success(mapOf("type" to "peerDiscovered", "data" to peerData))
            },
            onPayloadReceived = { peerId, bytes ->
                eventSink?.success(mapOf(
                    "type" to "payloadReceived",
                    "data" to mapOf("peerId" to peerId, "payload" to bytes)
                ))
            },
            onConnectionChanged = { peerId, connected ->
                eventSink?.success(mapOf(
                    "type" to if (connected) "peerConnected" else "peerDisconnected",
                    "data" to mapOf("peerId" to peerId)
                ))
            }
        )
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "initialize" -> {
                val peerId = call.argument<String>("peerId") ?: "unknown"
                bleTransport?.initialize(peerId)
                result.success(true)
            }
            "startDiscovery" -> {
                val success = bleTransport?.startDiscovery() ?: false
                result.success(success)
            }
            "stopDiscovery" -> {
                bleTransport?.stopDiscovery()
                result.success(true)
            }
            "send" -> {
                val peerId = call.argument<String>("peerId") ?: ""
                val payload = call.argument<ByteArray>("payload") ?: ByteArray(0)
                val success = bleTransport?.sendPayload(peerId, payload) ?: false
                result.success(success)
            }
            "broadcast" -> {
                val payload = call.argument<ByteArray>("payload") ?: ByteArray(0)
                val success = bleTransport?.broadcastPayload(payload) ?: false
                result.success(success)
            }
            "getCapabilities" -> {
                val bleCaps = bleTransport?.getCapabilities() ?: emptyMap<String, Any>()
                val wifiCaps = wifiTransport?.getCapabilities() ?: emptyMap<String, Any>()
                val perms = permissionManager?.getPermissionStatus() ?: emptyMap<String, Boolean>()
                val battery = batteryManager?.getBatteryInfo() ?: emptyMap<String, Any>()

                val combined = HashMap<String, Any>()
                combined.putAll(bleCaps)
                combined.putAll(wifiCaps)
                combined.putAll(perms)
                combined.putAll(battery)
                combined["platform"] = "android"

                result.success(combined)
            }
            "getBatteryInfo" -> {
                val info: Map<String, Any> = batteryManager?.getBatteryInfo() ?: emptyMap<String, Any>()
                result.success(info)
            }
            "getPermissions" -> {
                val perms: Map<String, Boolean> = permissionManager?.getPermissionStatus() ?: emptyMap<String, Boolean>()
                result.success(perms)
            }
            else -> result.notImplemented()
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        this.eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        this.eventSink = null
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        bleTransport?.shutdown()
        bleTransport = null
        context = null
    }
}

File: ios/Runner/MeshTransportPlugin.swift
import Flutter
import UIKit

public class MeshTransportPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
    private var eventSink: FlutterEventSink?
    private var multipeerTransport: MultipeerTransport?
    private var bluetoothTransport: BluetoothTransport?
    private var permissionManager: PermissionManager?
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let methodChannel = FlutterMethodChannel(name: "com.meshconnect/transport", binaryMessenger: registrar.messenger())
        let eventChannel = FlutterEventChannel(name: "com.meshconnect/events", binaryMessenger: registrar.messenger())
        
        let instance = MeshTransportPlugin()
        registrar.addMethodCallDelegate(instance, channel: methodChannel)
        eventChannel.setStreamHandler(instance)
        
        instance.setupTransports()
    }
    
    private func setupTransports() {
        permissionManager = PermissionManager()
        bluetoothTransport = BluetoothTransport()
        
        multipeerTransport = MultipeerTransport(
            onPeerDiscovered: { [weak self] peerData in
                self?.eventSink?(["type": "peerDiscovered", "data": peerData])
            },
            onPayloadReceived: { [weak self] peerId, data in
                self?.eventSink?([
                    "type": "payloadReceived",
                    "data": ["peerId": peerId, "payload": FlutterStandardTypedData(bytes: data)]
                ])
            },
            onConnectionChanged: { [weak self] peerId, connected in
                self?.eventSink?([
                    "type": connected ? "peerConnected" : "peerDisconnected",
                    "data": ["peerId": peerId]
                ])
            }
        )
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "initialize":
            let args = call.arguments as? [String: Any]
            let peerId = args?["peerId"] as? String ?? "ios-node"
            multipeerTransport?.initialize(peerIdString: peerId)
            result(true)
            
        case "startDiscovery":
            let success = multipeerTransport?.startDiscovery() ?? false
            result(success)
            
        case "stopDiscovery":
            multipeerTransport?.stopDiscovery()
            result(true)
            
        case "send":
            guard let args = call.arguments as? [String: Any],
                  let peerId = args["peerId"] as? String,
                  let typedData = args["payload"] as? FlutterStandardTypedData else {
                result(false)
                return
            }
            let success = multipeerTransport?.send(peerId: peerId, payload: typedData.data) ?? false
            result(success)
            
        case "broadcast":
            guard let args = call.arguments as? [String: Any],
                  let typedData = args["payload"] as? FlutterStandardTypedData else {
                result(false)
                return
            }
            let success = multipeerTransport?.broadcast(payload: typedData.data) ?? false
            result(success)
            
        case "getCapabilities":
            var caps = bluetoothTransport?.getCapabilities() ?? [:]
            caps["multipeerSupported"] = true
            caps["platform"] = "ios"
            if let perms = permissionManager?.getPermissionStatus() {
                for (k, v) in perms {
                    caps[k] = v
                }
            }
            UIDevice.current.isBatteryMonitoringEnabled = true
            let level = Int(UIDevice.current.batteryLevel * 100)
            caps["batteryLevel"] = level >= 0 ? level : 100
            caps["powerSaveMode"] = ProcessInfo.processInfo.isLowPowerModeEnabled
            result(caps)
            
        case "getBatteryInfo":
            UIDevice.current.isBatteryMonitoringEnabled = true
            let level = Int(UIDevice.current.batteryLevel * 100)
            result([
                "batteryLevel": level >= 0 ? level : 100,
                "powerSaveMode": ProcessInfo.processInfo.isLowPowerModeEnabled
            ])
            
        case "getPermissions":
            result(permissionManager?.getPermissionStatus() ?? [:])
            
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }
    
    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
}
